### PR TITLE
[ROX-11437] Batch resolve alerts operations to get performance boost

### PR DESCRIPTION
**Pull request by md2119**
## Description

- Added a method to mark bulk alerts resolved. 
- Unit tests compare resolving alerts in bulk vs resolving alerts individually.

Additionally:
- Do not perform read perm check when upserting records. If any read is being performed, it is only the side-effect of our business logic, but not expected by user
- Added write perm check to batch upsert. This will replace the service layer perm check.
- Fixed the unit test. The unit test was checking for wrong expectations. If the incoming context does not have correct permissions, an unauthorized access error should be returned, not a record not found error.

Benchmark results show significant improvement both on rocksdb (~100x) and postgres (~10x). 

```
# rocksdb:
BenchmarkDBs/rocksdb-8         	       1	2639426601 ns/op
BenchmarkDBs/markStale
BenchmarkDBs/markStale-8       	       1	231159944603 ns/op
BenchmarkDBs/markStaleBatch
BenchmarkDBs/markStaleBatch-8  	       1	2826356797 ns/op

# postgres:

BenchmarkDBsWithPostgres/markStale
BenchmarkDBsWithPostgres/markStale-8         	       1	13617595712 ns/op
BenchmarkDBsWithPostgres/markStaleBatch
BenchmarkDBsWithPostgres/markStaleBatch-8    	       1	1949332050 ns/op

PASS
```
## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
Unit tests

### CODE REVIEW

File: central/alert/datastore/datastore_impl.go [hunk 1]
```diff
@@ -187,22 +201,70 @@ func (ds *datastoreImpl) MarkAlertStale(ctx context.Context, id string) error {
...
+	idsAsBytes := make([][]byte, 0, len(ids))
+	for _, id := range ids {
+		idsAsBytes = append(idsAsBytes, []byte(id))
+	}
+
+	var resolvableAlerts []*storage.Alert
+	err := ds.keyFence.DoStatusWithLock(dackboxConcurrency.DiscreteKeySet(idsAsBytes...), func() error {
+		var err error
+		var missing []int
+		// Avoid `ds.GetMany` since it performs read SAC check, which is not expected for user performing an upsert.
```
Comment (theencee) [inline central/alert/datastore/datastore_impl.go:238]: [nit] comment is incorrect

File: central/alert/datastore/datastore_impl.go [hunk 2]
```diff
@@ -187,22 +201,70 @@ func (ds *datastoreImpl) MarkAlertStale(ctx context.Context, id string) error {
...
+			ok, err := alertSAC.WriteAllowed(ctx, sacKeyForAlert(alert)...)
+			if err != nil {
+				return err
+			}
+			if !ok {
+				return sac.ErrResourceAccessDenied
+			}
+		}
+
+		for _, alert := range resolvableAlerts {
```
Comment (theencee) [inline central/alert/datastore/datastore_impl.go:259]: [nit] combine with for block above?

File: central/alert/datastore/datastore_impl.go [hunk 3]
```diff
@@ -187,22 +201,70 @@ func (ds *datastoreImpl) MarkAlertStale(ctx context.Context, id string) error {
...
+	}
+	if !ok {
 		return sac.ErrResourceAccessDenied
 	}
 	alert.State = storage.ViolationState_RESOLVED
 	alert.ResolvedAt = types.TimestampNow()
 	return ds.updateAlertNoLock(ctx, alert)
 }
 
+func (ds *datastoreImpl) MarkAlertStaleBatch(ctx context.Context, ids ...string) ([]*storage.Alert, error) {
```
[the remark this answers is not in the document]
Comment (md2119) [reply central/alert/datastore/datastore_impl.go:226]: I know, it is confusing. The update func IMO are not really batches, but split into individual go routines. Therefore at the DB level, they still wait on locks.

[the remark this answers is not in the document]
Comment (md2119) [reply central/alert/datastore/datastore_impl.go:226]: I will be focusing on marking alerts as stale workflow in this PR.

File: central/alert/datastore/datastore_impl.go [hunk 4]
```diff
@@ -141,13 +143,25 @@ func (ds *datastoreImpl) UpdateAlertBatch(ctx context.Context, alert *storage.Al
 	defer waitGroup.Done()
 	ds.keyedMutex.Lock(alert.GetId())
 	defer ds.keyedMutex.Unlock(alert.GetId())
-	oldAlert, exists, err := ds.GetAlert(ctx, alert.GetId())
+	// Avoid `ds.GetMany` since it performs read SAC check, which is not expected for user performing an upsert.
```
[the remark this answers is not in the document]
Comment (md2119) [reply central/alert/datastore/datastore_impl.go:146]: Ah, yes!

**Review by theencee — APPROVED** [review]

**Pull request merged**